File: Jonghoon Jang/2023-03-16-maximum-product-subarray.py
"""
70. Climbing Stairs
"""

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curr_min = 1
        curr_max = 1

        for num in nums:
            temp = curr_max * num
            curr_max = max(curr_max * num, curr_min * num, num)  # [-1, 8] itself
            curr_min = min(temp, curr_min * num, num)
            res = max(res, curr_max)
        return res
def main():
    s = Solution()


if __name__ == "__main__":
    main()

File: MinjiJu/2023-03-16-Maximum-Product-Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int res=nums[0], left=0, right=0;
        
        for(int i=0; i<n; i++){
            left = (left ? left : 1)*nums[i];
            right = (right ? right : 1)*nums[n-1-i];
            res = max(res, max(left,right));
        }
        return res;
    }
};